src/main/java/de/zalando/aruha/nakadi/service/subscription/StreamingContext.java

```diff
@@ -0,0 +1,170 @@
+
+    StreamingContext(
+            final SubscriptionOutput out,
+            final StreamParameters parameters,
+            final Session session,
+            final ScheduledExecutorService timer,
+            final ZkSubscriptionClient zkClient,
+            final KafkaClient kafkaClient,
+            final BiFunction<Session[], Partition[], Partition[]> rebalancer,
+            final long kafkaPollTimeout) {
+        this.out = out;
+        this.parameters = parameters;
+        this.session = session;
+        this.rebalancer = rebalancer;
+        this.timer = timer;
+        this.zkClient = zkClient;
+        this.kafkaClient = kafkaClient;
+        this.kafkaPollTimeout = kafkaPollTimeout;
+    }
+
+    public StreamParameters getParameters() {
+        return parameters;
+    }
+
+    public ZkSubscriptionClient getZkClient() {
+        return zkClient;
+    }
+
+    public String getSessionId() {
+        return session.getId();
+    }
+
+    public KafkaClient getKafkaClient() {
+        return kafkaClient;
+    }
+
+    public SubscriptionOutput getOut() {
+        return out;
+    }
+
+    public long getKafkaPollTimeout() {
+        return kafkaPollTimeout;
+    }
+
+    @Override
+    public void stream() throws InterruptedException {
+        currentState = new State() {
+            @Override
+            public void onEnter() {
+            }
+        };
+
+        // Add first task - switch to starting state.
+        switchState(new StartingState());
+
+        while (null != currentState) {
+            final Runnable task = taskQueue.poll(1, TimeUnit.MINUTES);
+            try {
+                task.run();
+            } catch (final SubscriptionWrappedException ex) {
+                LOG.error("Failed to process task " + task + ", will rethrow original error", ex);
+                switchState(new CleanupState(ex.getSourceException()));
+            } catch (final RuntimeException ex) {
+                LOG.error("Failed to process task " + task + ", code carefully!", ex);
+                switchState(new CleanupState(ex));
+            }
+        }
+    }
+
+    public void switchState(final State newState) {
+        this.addTask(() -> {
+            if (currentState != null) {
+                LOG.info("Switching state from " + currentState.getClass().getSimpleName());
+                currentState.onExit();
+            } else {
+                LOG.info("Connection died, no new state switches available");
+                return;
+            }
+            currentState = newState;
+            if (null != currentState) {
+                LOG.info("Switching state to " + currentState.getClass().getSimpleName());
+                currentState.setContext(this);
+                currentState.onEnter();
+            } else {
+                LOG.info("No next state found, dying");
+            }
+        });
+    }
+
+    public void registerSession() {
```

Comment:
It's misleading to have a method called register session that also triggers rebalance.